scripts/build-vm-image.sh

```diff
@@ -16,26 +16,41 @@
 set -eu
 
 # Ensure all required env vars are supplied.
-for var in BUCKET CHEF_DIR KEY_FILE_PATH PACKER_BINARY PACKER_DIR PROJECT SOLUTION_NAME; do
+for var in BUCKET CHEF_DIR PACKER_BINARY PACKER_DIR PROJECT SOLUTION_NAME; do
   if ! [[ -v "${var}" ]]; then
     echo "${var} env variable is required"
     exit 1
   fi
 done
 
+if [[ ! -v KEY_FILE_PATH ]] && [[ ! -d "${HOME}/.config/gcloud" ]]; then
+  echo "Either the KEY_FILE_PATH must be set or ~/.config/gcloud must be mounted."
+  exit 1
+fi
+
+if [[ -d "${HOME}/.config/gcloud" ]] && [[ ! -v SERVICE_ACCOUNT_EMAIL ]]; then
+  echo "It appears you have mounted local credentials.  Therefore a SERVICE_ACCOUNT_EMAIL must be specified."
+  exit 1
+fi
+
 function _register_gcloud_config() {
-  local -r config_name="imagebuilder${RANDOM}"
+  
+  # If a service account e-mail is not specified, assume a KEY_FILE is being used.
+  if [[ ! -v SERVICE_ACCOUNT_EMAIL ]]; then
```

Comment:
This checks for SERVICE_ACCOUNT_EMAIL but makes use of KEY_FILE_PATH. It is not clear that KEY_FILE_PATH will be available if SERVICE_ACCOUNT_EMAIL is not.

It would be more clear if the checks above were like

if SERVICE_ACCOUNT_EMAIL:
  assert everything necessary for service account
else:
  assert KEY_FILE_PATH is present

Reply:
Looking at the whole block, looks like it is easier to read if you invoke _register_gcloud_config inside the blocks as well.

It is generally nice to have flat checks at the beginning, but in this case we need to coordinate 3 different switches to understand what is happening.

Reply:
Thanks for the review.  Makes sense.